File: PayPal/src/test/java/com/braintreepayments/api/PayPalVaultRequestUnitTest.java
package com.braintreepayments.api;

import android.content.pm.Signature;
import android.os.Parcel;

import org.json.JSONException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.mock;

@RunWith(RobolectricTestRunner.class)
public class PayPalVaultRequestUnitTest {

    @Test
    public void newPayPalVaultRequest_setsDefaultValues() {
        PayPalVaultRequest request = new PayPalVaultRequest(false);

        assertNull(request.getLocaleCode());
        assertFalse(request.isShippingAddressRequired());
        assertNull(request.getShippingAddressOverride());
        assertNull(request.getDisplayName());
        assertNull(request.getLandingPageType());
        assertFalse(request.getShouldOfferCredit());
        assertFalse(request.hasUserLocationConsent());
    }

    @Test
    public void newPayPalVaultRequest_without_hasUserLocationConsent_defaults_to_false() {
        PayPalVaultRequest request = new PayPalVaultRequest();
        assertFalse(request.hasUserLocationConsent());
    }

    @Test
    public void setsValuesCorrectly() {
        PostalAddress postalAddress = new PostalAddress();
        PayPalVaultRequest request = new PayPalVaultRequest(true);
        request.setLocaleCode("US");
        request.setBillingAgreementDescription("Billing Agreement Description");
        request.setShippingAddressRequired(true);
        request.setShippingAddressOverride(postalAddress);
        request.setDisplayName("Display Name");
        request.setRiskCorrelationId("123-correlation");
        request.setLandingPageType(PayPalRequest.LANDING_PAGE_TYPE_LOGIN);
        request.setShouldOfferCredit(true);

        assertEquals("US", request.getLocaleCode());
        assertEquals("Billing Agreement Description", request.getBillingAgreementDescription());
        assertTrue(request.isShippingAddressRequired());
        assertEquals(postalAddress, request.getShippingAddressOverride());
        assertEquals("Display Name", request.getDisplayName());
        assertEquals("123-correlation", request.getRiskCorrelationId());
        assertEquals(PayPalRequest.LANDING_PAGE_TYPE_LOGIN, request.getLandingPageType());
        assertTrue(request.getShouldOfferCredit());
        assertTrue(request.hasUserLocationConsent());
    }

    @Test
    public void parcelsCorrectly() {
        PayPalVaultRequest request = new PayPalVaultRequest(true);
        request.setLocaleCode("en-US");
        request.setBillingAgreementDescription("Billing Agreement Description");
        request.setShippingAddressRequired(true);
        request.setShippingAddressEditable(true);
        request.setShouldOfferCredit(true);

        PostalAddress postalAddress = new PostalAddress();
        postalAddress.setRecipientName("Postal Address");
        request.setShippingAddressOverride(postalAddress);

        request.setLandingPageType(PayPalRequest.LANDING_PAGE_TYPE_LOGIN);
        request.setDisplayName("Display Name");
        request.setRiskCorrelationId("123-correlation");
        request.setMerchantAccountId("merchant_account_id");

        ArrayList<PayPalLineItem> lineItems = new ArrayList<>();
        lineItems.add(new PayPalLineItem(PayPalLineItem.KIND_DEBIT, "An Item", "1", "1"));
        request.setLineItems(lineItems);

        Parcel parcel = Parcel.obtain();
        request.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);
        PayPalVaultRequest result = PayPalVaultRequest.CREATOR.createFromParcel(parcel);

        assertEquals("en-US", result.getLocaleCode());
        assertEquals("Billing Agreement Description",
            result.getBillingAgreementDescription());
        assertTrue(result.getShouldOfferCredit());
        assertTrue(result.isShippingAddressRequired());
        assertTrue(result.isShippingAddressEditable());
        assertEquals("Postal Address", result.getShippingAddressOverride()
            .getRecipientName());
        assertEquals(PayPalRequest.LANDING_PAGE_TYPE_LOGIN, result.getLandingPageType());
        assertEquals("Display Name", result.getDisplayName());
        assertEquals("123-correlation", result.getRiskCorrelationId());
        assertEquals("merchant_account_id", result.getMerchantAccountId());
        assertEquals(1, result.getLineItems().size());
        assertEquals("An Item", result.getLineItems().get(0).getName());
        assertTrue(result.hasUserLocationConsent());
    }

    @Test
    public void createRequestBody_sets_userAuthenticationEmail_when_not_null() throws JSONException {
        String payerEmail = "payer_email@example.com";
        PayPalVaultRequest request = new PayPalVaultRequest();

        request.setUserAuthenticationEmail(payerEmail);
        String requestBody = request.createRequestBody(
            mock(Configuration.class),
            mock(Authorization.class),
            "success_url",
            "cancel_url"
        );

        assertTrue(requestBody.contains("\"payer_email\":" + "\"" + payerEmail + "\""));
    }

    @Test
    public void createRequestBody_sets_userPhoneNumber_when_not_null() throws JSONException {
        String userPhoneNumber = "1231231234";
        PayPalVaultRequest request = new PayPalVaultRequest(true);

        request.setUserPhoneNumber(userPhoneNumber);
        String requestBody = request.createRequestBody(
            mock(Configuration.class),
            mock(Authorization.class),
            "success_url",
            "cancel_url"
        );

        assertTrue(requestBody.contains("\"phone\":" + "\"" + userPhoneNumber + "\""));
    }

    @Test
    public void createRequestBody_does_not_set_userPhoneNumber_when_email_is_empty() throws JSONException {
        String userPhoneNumber = "";
        PayPalVaultRequest request = new PayPalVaultRequest(true);

        request.setUserPhoneNumber(userPhoneNumber);
        String requestBody = request.createRequestBody(
            mock(Configuration.class),
            mock(Authorization.class),
            "success_url",
            "cancel_url"
        );

        assertFalse(requestBody.contains("\"phone\":" + "\"" + userPhoneNumber + "\""));
    }
}

File: PayPal/src/test/java/com/braintreepayments/api/PayPalVaultRequestUnitTest.kt
package com.braintreepayments.api

import android.os.Parcel
import junit.framework.TestCase.*
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.*

@RunWith(RobolectricTestRunner::class)
class PayPalVaultRequestUnitTest {

    @Test
    fun newPayPalVaultRequest_setsDefaultValues() {
        val request = PayPalVaultRequest()

        assertNull(request.localeCode)
        assertFalse(request.isShippingAddressRequired)
        assertNull(request.shippingAddressOverride)
        assertNull(request.displayName)
        assertNull(request.landingPageType)
        assertFalse(request.shouldOfferCredit)
    }

    @Test
    fun setsValuesCorrectly() {
        val postalAddress = PostalAddress()
        val request = PayPalVaultRequest()

        request.localeCode = "US"
        request.billingAgreementDescription = "Billing Agreement Description"
        request.isShippingAddressRequired = true
        request.shippingAddressOverride = postalAddress
        request.displayName = "Display Name"
        request.landingPageType = PayPalRequest.LANDING_PAGE_TYPE_LOGIN
        request.shouldOfferCredit = true
        assertEquals("US", request.localeCode)
        assertEquals("Billing Agreement Description", request.billingAgreementDescription)
        assertTrue(request.isShippingAddressRequired)
        assertEquals(postalAddress, request.shippingAddressOverride)
        assertEquals("Display Name", request.displayName)
        assertEquals(PayPalRequest.LANDING_PAGE_TYPE_LOGIN, request.landingPageType)
        assertTrue(request.shouldOfferCredit)
    }

    @Test
    fun parcelsCorrectly() {
        val request = PayPalVaultRequest()
        request.localeCode = "en-US"
        request.billingAgreementDescription = "Billing Agreement Description"
        request.isShippingAddressRequired = true
        request.isShippingAddressEditable = true
        request.shouldOfferCredit = true

        val postalAddress = PostalAddress()
        postalAddress.recipientName = "Postal Address"
        request.shippingAddressOverride = postalAddress
        request.landingPageType = PayPalRequest.LANDING_PAGE_TYPE_LOGIN
        request.displayName = "Display Name"
        request.merchantAccountId = "merchant_account_id"

        val lineItems = ArrayList<PayPalLineItem>()
        lineItems.add(PayPalLineItem(PayPalLineItem.KIND_DEBIT, "An Item", "1", "1"))
        request.setLineItems(lineItems)

        val parcel = Parcel.obtain()
        request.writeToParcel(parcel, 0)
        parcel.setDataPosition(0)
        val result = PayPalVaultRequest.CREATOR.createFromParcel(parcel)

        assertEquals("en-US", result.localeCode)
        assertEquals("Billing Agreement Description",
                result.billingAgreementDescription)
        assertTrue(result.shouldOfferCredit)
        assertTrue(result.isShippingAddressRequired)
        assertTrue(result.isShippingAddressEditable)
        assertEquals("Postal Address", result.shippingAddressOverride?.recipientName)
        assertEquals(PayPalRequest.LANDING_PAGE_TYPE_LOGIN, result.landingPageType)
        assertEquals("Display Name", result.displayName)
        assertEquals("merchant_account_id", result.merchantAccountId)
        assertEquals(1, result.lineItems.size)
        assertEquals("An Item", result.lineItems[0].name)
    }
}